### PR TITLE
ODH install using ArgoCD

### PR DESCRIPTION
**Pull request by martinpovolny**
based on @anishasthana 's demo

preview here: https://github.com/martinpovolny/continuous-deployment/blob/odh-install/docs/downstream/odh-install-crc.md

~to be rebased after https://github.com/operate-first/continuous-deployment/pull/33 is merged~

### CODE REVIEW

**Comment by martinpovolny** [comment]
Issues:

* the superset route lacks HTTPS

File: docs/downstream/odh-install-crc.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+Go to https://argocd-server-aicoe-argocd-dev.apps-crc.testing/applications and click "New App". Enter these values:
+
+| Field            | Value                                                |
+|------------------|------------------------------------------------------|
+| Project:         | Default                                              |
+| Cluster:         | dev-cluster (https://api.crc.testing.6443)           |
+| Namespace:       | odh-operator                                         |
+| Repo URL:        | https://github.com/anishasthana/my-odh-install.git   |
+| Target revision: | HEAD                                                 |
+| Path:            | odh-operator/base                                    |
```
Comment (tumido) [inline docs/downstream/odh-install-crc.md]: I think it would be better to provide the `Application` yaml snippet here, instead of a table of what user should click and type in the UI. What do you think?

Swap it for something like:

```
cat <<EOF | oc apply -f -
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: odh
spec:
  destination:
    namespace: odh-operator
    server: https://api.crc.testing.6443
  project: default
  source:
    path: odh-operator/base
    repoURL: https://github.com/anishasthana/my-odh-install.git
    targetRevision: HEAD
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
    syncOptions:
      - Validate=false
EOF
```

**Review by tumido — CHANGES_REQUESTED** [review]
I think providing the `Application` manifests directly would make it much easier to follow.

File: docs/downstream/odh-install-crc.md (hunk 1 above)
Comment (martinpovolny) [reply docs/downstream/odh-install-crc.md]: yes, it makes sense

**Comment by martinpovolny** [comment]
Also we should probably fork @anishasthana 's repo somewhere. Under operate-first probably?

And the instructions should probably mention people having to fork the repo and deploy from their own fork. WDYT?

**Comment by anishasthana** [comment]
> Also we should probably fork @anishasthana 's repo somewhere. Under operate-first probably?
> 
> And the instructions should probably mention people having to fork the repo and deploy from their own fork. WDYT?

I agree, I think we can create a brand new repo somewhere within the operate-first umbrella.

File: docs/downstream/odh-install-crc.md (hunk 1 above)
Comment (martinpovolny) [reply docs/downstream/odh-install-crc.md]: But I'd rather put the resources into the repo than have them in-line in the doc. It will be easier to (automatically) test if things keep working. WDYT?

File: docs/downstream/crc.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Deployment on CRC
+
+These is how to deploy ArgoCD on [CRC](https://developers.redhat.com/products/codeready-containers/overview).
```
Comment (durandom) [inline docs/downstream/crc.md]: `This is`...

File: docs/downstream/odh-install-crc.md (hunk 1 above)
Comment (durandom) [reply docs/downstream/odh-install-crc.md]: +1 on the yaml snippet.

Put the yaml into the repo here in `/docs/downstream` for now. 
In the future I think the ODH specific steps should be in a separate repo anyways

File: examples/crc-create-myadmin.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
+oc create user myadmin
```
Comment (durandom) [inline examples/crc-create-myadmin.sh]: Let's move this to `docs/downstream` as well

**Review by durandom — APPROVED** [review]
I think we can merge this as is. 

Follow up

* recreate https://github.com/anishasthana/my-odh-install in our org
* exchange the table view and where to click in the UI with yaml files
* move the ODH specific parts to an ODH specific repo

File: docs/downstream/crc.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+## Instalation Steps
```
Comment (HumairAK) [inline docs/downstream/crc.md]: Installation*

File: docs/downstream/crc.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+   `crc start`
+
+ * Get `kustomize` and KSOPS using steps in [manage_your_app_secrets.md](../manage_your_app_secrets.md)
+
+   * If you hit:\
+``` unrecognized import path "vbom.ml/util": https fetch: Get "https://vbom.ml/util?go-get=1": dial tcp ```\
+fix it with https://github.com/viaduct-ai/kustomize-sops/issues/60
+
+* As an alternative to installing the prerequisites locally, you can use Toolbox: https://github.com/containers/toolbox \
+   `toolbox create --image quay.io/aicoe/of-toolbox:v0.1.0` \
```
Comment (HumairAK) [inline docs/downstream/crc.md]: I would remove the instructions here on how to use the toolbox since they are mentioned in the link we'd need to make changes to multiple files if the versions change otherwise :)

File: docs/downstream/crc.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
+# Deployment on CRC
+
+These is how to deploy ArgoCD on [CRC](https://developers.redhat.com/products/codeready-containers/overview).
+
+## Instalation Steps
+
+ * Setup CRC https://developers.redhat.com/products/codeready-containers/overview
```
Comment (HumairAK) [inline docs/downstream/crc.md:7]: Can you convert all your links to hyperlinks  in the form [ text ] ( link )  it's a lot less cluttered this way imo

File: docs/downstream/crc.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+   Then you have all the tools needed running in a separate container.
+
+ * Fork https://github.com/operate-first/continuous-deployment
+
+ * Import GPG key that is used by kustomize to encrypt the secrets.\
+	```base64 -d < examples/key.asc | gpg --import ```
+
+ * Proceed with [setup_argocd_dev_environment.md](../setup_argocd_dev_environment.md)
+
+Your ArgoCD instance will be running at https://argocd-server-aicoe-argocd-dev.apps-crc.testing/
```
Comment (HumairAK) [inline docs/downstream/crc.md]: I would mention that they can login using openshift auth or as admin (user/pass: admin/admin)

File: docs/downstream/odh-install-crc.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+## Preparation
+
+First make sure that you have the OpenDataHub operator available in your OpenShift.
+
+```bash
+oc get packagemanifests -n openshift-marketplace | grep opendatahub-operator
+```
+
+If it's not present, then you need to troubleshoot the marketplace: https://github.com/operator-framework/operator-marketplace/issues/344
```
Comment (HumairAK) [inline docs/downstream/odh-install-crc.md:11]: Can we convert these links to hyperlinks as well  in the form [ text ] ( link )

File: docs/downstream/odh-install-crc.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+
+Then we need to create a projects named `odh-operator` and `odh-deployment`. (The name of the project is hard-coded in the `kustomize` files in https://github.com/anishasthana/my-odh-install and you can change is there using `kustomize`.)
```
Comment (HumairAK) [inline docs/downstream/odh-install-crc.md:14]: > Then we need to create a projects named 
project * 

> https://github.com/anishasthana/my-odh-install 

This is @anishasthana 's personal repo, are we okay with linking there? is there a more official location? If not, then maybe we bring these files in to this org?

File: docs/downstream/odh-install-crc.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+```bash
+oc new-project odh-operator
+oc new-project odh-deployment
+```
+
+Our ArgoCD SA `argocd-manager` needs to have access to the new project. On CRC I suggest giving `argocd-manager` access to all projects by creating a ClusterRoleBinding to a ClusterRole.
+
+`oc apply -f examples/argocd-cluster-binding.yaml`
```
Comment (HumairAK) [inline docs/downstream/odh-install-crc.md:23]: I don't see this file?

File: docs/downstream/odh-install-crc.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+```bash
+oc patch secret dev-cluster-spec -n aicoe-argocd-dev --type='json' -p="[{'op': 'add', 'path': '/data/namespace', 'value':''}]"
+```
+
+Be aware that if you change this value in the ArgoCD UI, you might loose the stored credentials for the cluster due to a bug in ArgoCD.
+
+## Creating the ArgoCD applications
+
+Now we can proceed with creating the "applications" in ArgoCD. We will create 2 applications:
```
Comment (HumairAK) [inline docs/downstream/odh-install-crc.md]: I would rephrase this: 
> Now we can proceed with creating the "applications" in ArgoCD 

to say: 

> Now we can proceed with creating the ArgoCD Applications.

File: docs/downstream/odh-install-crc.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+  1. For the ODH operator. (Installs the ODH operator itself `odh-operator`.)
+  1. For the ODH deployment. (Installs ODH components into `odh-deployment`.)
```
Comment (HumairAK) [inline docs/downstream/odh-install-crc.md]: Remove "For" and just say "The ODH operator" and "The ODH deployment"

File: docs/downstream/odh-install-crc.md [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+| Namespace:       | odh-operator                                         |
+| Repo URL:        | https://github.com/anishasthana/my-odh-install.git   |
+| Target revision: | HEAD                                                 |
+| Path:            | odh-operator/base                                    |
+
+This creates an app from definition in the repo under the path `odh-operator/base` and deploys it to the `odh-operator` namespace on the `dev-cluster`. Please note that the namespace is also hard-coded in the repo so changing it requires changing files in the repo.
+
+
+Create the app and you will see Argo deploying resources:
+(doc/assets/images/upstream/odh-operator-argocd.png)
```
Comment (HumairAK) [inline docs/downstream/odh-install-crc.md]: this link is invalid, I'm guessing it's because it should be "docs" not "doc"

File: examples/crc-create-myadmin.sh (hunk 3 above)
Comment (HumairAK) [reply examples/crc-create-myadmin.sh]: In that case I'd suggest putting this in a markdown, it would be weird having a script in a docs folder :)

**Review by HumairAK — CHANGES_REQUESTED** [review]

File: docs/downstream/crc.md (hunk 2 above)
Comment (martinpovolny) [reply docs/downstream/crc.md]: Fixed in another PR.

File: examples/crc-create-myadmin.sh (hunk 3 above)
Comment (martinpovolny) [reply examples/crc-create-myadmin.sh]: I'd rather put it to `examples/`. But can create `examples/upstream/crc` or `examples/crc` WDTY?

It probably does not matter much at this point and we can figure it up as we add more content anyway.

File: docs/downstream/crc.md (hunk 4 above)
Comment (martinpovolny) [reply docs/downstream/crc.md]: fixing...

File: docs/downstream/crc.md (hunk 5 above)
Comment (martinpovolny) [reply docs/downstream/crc.md]: This was part of another PR and is here because of the branching. Let's deal with it separately if needed.

File: docs/downstream/crc.md (hunk 6 above)
Comment (martinpovolny) [reply docs/downstream/crc.md:7]: I do it for internal links. For external links I keep the URL in text. I personally like to see where I am heading when I click links. But it's probably a matter of preference.

File: docs/downstream/crc.md (hunk 7 above)
Comment (martinpovolny) [reply docs/downstream/crc.md]: +1

File: docs/downstream/odh-install-crc.md (hunk 8 above)
Comment (martinpovolny) [reply docs/downstream/odh-install-crc.md:11]: Same as above. I suggest doing it for relative links and keeping the external ones.

File: docs/downstream/odh-install-crc.md (hunk 10 above)
Comment (martinpovolny) [reply docs/downstream/odh-install-crc.md:23]: added

File: docs/downstream/odh-install-crc.md (hunk 11 above)
Comment (martinpovolny) [reply docs/downstream/odh-install-crc.md]: +1

File: docs/downstream/odh-install-crc.md (hunk 12 above)
Comment (martinpovolny) [reply docs/downstream/odh-install-crc.md]: +1

File: docs/downstream/odh-install-crc.md (hunk 13 above)
Comment (martinpovolny) [reply docs/downstream/odh-install-crc.md]: thx, it was a line I forgot to remove. the correct URL is one line down

File: docs/downstream/crc.md (hunk 7 above)
Comment (martinpovolny) [reply docs/downstream/crc.md]: It was it the other PR, but adding it anyway.

File: docs/downstream/odh-install-crc.md (hunk 9 above)
Comment (anishasthana) [reply docs/downstream/odh-install-crc.md:14]: I vote we create a new repo within the operatre first umbrella @HumairAK

Comment (durandom) [reply docs/downstream/odh-install-crc.md:14]: @martinpovolny I just copied it to https://github.com/operate-first/odh

**Comment by martinpovolny** [comment]
> recreate https://github.com/anishasthana/my-odh-install in our org

done by you, I updated the links

> exchange the table view and where to click in the UI with yaml files

I have added links to the files and the files while keeping the tables. I think that it helps to understand what is happening to see the tables.

**Comment by martinpovolny** [comment]
@tumido : Can you merge this one now or do you have other comments?

**Review by tumido — APPROVED** [review]
